api/turing/api/ensemblers_api.go

```diff
@@ -109,18 +111,248 @@ func (c EnsemblersController) UpdateEnsembler(
+	}
+	if len(activeRouter) >= 1 {
+		return http.StatusBadRequest, fmt.Errorf("there are active router version using this ensembler")
+	}
+	return http.StatusOK, nil
+}
+
+func (c EnsemblersController) checkCurrentRouterVersion(options EnsemblersPathOptions) (int, error) {
+	activeOption := service.RouterVersionListOptions{
+		ProjectID:   options.ProjectID,
+		EnsemblerID: options.EnsemblerID,
+		IsCurrent:   true,
+	}
+	activeRouter, err := c.RouterVersionsService.ListRouterVersionsWithFilter(activeOption)
+	if err != nil {
+		return http.StatusInternalServerError, err
+	}
+	if len(activeRouter) >= 1 {
+		return http.StatusBadRequest, fmt.Errorf("" +
+			"there are router version that is currently being used by a router using this ensembler",
+		)
+	}
+	return http.StatusOK, nil
+}
+
+func (c EnsemblersController) checkActiveEnsemblingJob(options EnsemblersPathOptions) (int, error) {
+	ensemblingJobActiveOption := service.EnsemblingJobListOptions{
+		EnsemblerID: options.EnsemblerID,
+		Statuses: []models.Status{
+			models.JobPending,
+			models.JobBuildingImage,
+			models.JobRunning,
+		},
+	}
+	activeEnsemblingJobs, err := c.EnsemblingJobService.List(ensemblingJobActiveOption)
+	if err != nil {
+		return http.StatusInternalServerError, err
+	}
+	if activeEnsemblingJobs.Paging.Total >= 1 {
+		return http.StatusBadRequest, fmt.Errorf("there are active ensembling job using this ensembler")
+	}
+	return http.StatusOK, nil
+}
+func (c EnsemblersController) deleteInactiveRouterVersion(options EnsemblersPathOptions) (int, error) {
+	routerVersionStatusInactive := []models.RouterVersionStatus{
+		models.RouterVersionStatusFailed,
+		models.RouterVersionStatusUndeployed,
+	}
+	inactiveOption := service.RouterVersionListOptions{
+		ProjectID:   options.ProjectID,
+		EnsemblerID: options.EnsemblerID,
+		Statuses:    routerVersionStatusInactive,
+	}
+	inactiveRouter, err := c.RouterVersionsService.ListRouterVersionsWithFilter(inactiveOption)
+	if err != nil {
+		return http.StatusInternalServerError, err
+	}
+
+	for _, routerVersion := range inactiveRouter {
+		err = c.RouterVersionsService.Delete(routerVersion)
+		if err != nil {
+			return http.StatusInternalServerError, err
+		}
+	}
+	return http.StatusOK, nil
+}
+func (c EnsemblersController) deleteInactiveEnsemblingJob(options EnsemblersPathOptions) (int, error) {
+	ensemblingJobInactiveOption := service.EnsemblingJobListOptions{
+		EnsemblerID: options.EnsemblerID,
+		Statuses: []models.Status{
+			models.JobFailed,
+			models.JobCompleted,
+			models.JobFailedBuildImage,
+			models.JobFailedSubmission,
+		},
+	}
+	inactiveEnsemblingJobs, err := c.EnsemblingJobService.List(ensemblingJobInactiveOption)
+	if err != nil {
+		return http.StatusInternalServerError, err
+	}
+	if inactiveEnsemblingJobs.Paging.Total > 0 {
+		if results, ok := inactiveEnsemblingJobs.Results.([]*models.EnsemblingJob); ok {
+			for _, ensemblingJob := range results {
+				if ensemblingJob.Status == models.JobFailed {
```

Comment:
wait i have a reason to do this, but let me check first to confirm the issue, i think it was mainly because the failed one doesn't need to be cleaned up again from the spark, but other error need cleanup action on the spark

Reply:
im not sure about the jobDeletion Spark Application Status, but after seeing the implementation of the code 
```go
func (c *ensemblingController) Delete(namespace string, ensemblingJob *models.EnsemblingJob) error {
	sa, err := c.clusterController.GetSparkApplication(context.Background(), namespace, ensemblingJob.Name)
	if err != nil {
		c.cleanup(ensemblingJob.Name, namespace)
		// Not found, we do not consider this as an error because its just no further
		// action required on this part.
		return nil
	}
	c.cleanup(ensemblingJob.Name, namespace)
	return c.clusterController.DeleteSparkApplication(context.Background(), namespace, sa.Name)
}
```

if we mark all to terminating, its ok since if the job doesnt exist, it wont show any error, and it will have the cleanup if the job have a spark application / image building job on kaniko

So if we just do MarkEnsemblingJobForTermination for all job deletion handling, its ok, since they will handle if the job not found from kaniko/spark. If theres no data on kaniko/spark they will just to try delete the data from database.